das/syncing_fallback_storage.go

```diff
@@ -212,31 +217,45 @@ func (s *l1SyncService) processBatchDelivered(ctx context.Context, batchDelivere
 	binary.BigEndian.PutUint64(header[32:40], deliveredEvent.AfterDelayedMessagesRead.Uint64())
 
 	data = append(header, data...)
-	preimages := make(map[arbutil.PreimageType]map[common.Hash][]byte)
-	preimageRecorder := daprovider.RecordPreimagesTo(preimages)
-	if _, err = daprovider.RecoverPayloadFromDasBatch(ctx, deliveredEvent.BatchSequenceNumber.Uint64(), data, s.dataSource, preimageRecorder, true); err != nil {
+	var payload []byte
+	if payload, err = daprovider.RecoverPayloadFromDasBatch(ctx, deliveredEvent.BatchSequenceNumber.Uint64(), data, s.dataSource, nil, true); err != nil {
 		if errors.Is(err, daprovider.ErrSeqMsgValidation) {
 			log.Error(err.Error())
+			// TODO why is this just logged and not returned?
 		} else {
 			log.Error("recover payload failed", "txhash", batchDeliveredLog.TxHash, "data", data)
 			return err
 		}
 	}
-	for _, preimages := range preimages {
-		for hash, contents := range preimages {
-			var err error
-			if s.config.CheckAlreadyExists {
-				_, err = s.syncTo.GetByHash(ctx, hash)
-			}
-			if err == nil || errors.Is(err, ErrNotFound) {
-				if err := s.syncTo.Put(ctx, contents, storeUntil); err != nil {
+
+	if payload != nil {
+		var skip bool
+		if s.config.CheckAlreadyExists {
+			// The DA cert in the sequencer data may use the V0 style flat hash, but
+			// Put always uses the V1 style dastree Hash, so just check for the
+			// existence of the batch by dastree Hash.
+			dataHash := dastree.Hash(payload)
+			existingPayload, err := s.syncTo.GetByHash(ctx, dataHash)
+			if err != nil {
+				if !errors.Is(err, ErrNotFound) {
 					return err
 				}
+			} else if !bytes.Equal(existingPayload, payload) {
+				log.Error("mismatch between existing and retrieved data in l1SyncService", "existing", pretty.PrettyBytes(existingPayload), "retrieved", pretty.PrettyBytes(payload), "dataHash", dataHash)
+				return errors.New("mismatch between existing and retrived data in l1SyncService")
 			} else {
+				log.Info("l1SyncService skipping syncing existing data for", "dataHash", dataHash)
+				skip = true
+			}
+		}
+
+		if !skip {
```

Comment:
as discussed - remove the skip logic in case there are duplicate payloads

Reply:
Thanks for noticing this issue!